hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java

```diff
@@ -584,16 +573,36 @@ private Pair<SchemaProvider, Pair<String, JavaRDD<HoodieRecord>>> fetchFromSourc
                 (SchemaProvider) new DelegatingSchemaProvider(props, hoodieSparkContext.jsc(), dataAndCheckpoint.getSchemaProvider(),
                     new SimpleSchemaProvider(hoodieSparkContext.jsc(), targetSchema, props)))
             .orElse(dataAndCheckpoint.getSchemaProvider());
+      }
+
+      // Short circuit for bulk insert to gain performance benefits of row writing
+      if (recordType == HoodieRecordType.SPARK) {
+        rowDatasetOptional = transformed;
+        noNewData = (!rowDatasetOptional.isPresent()) || (rowDatasetOptional.get().isEmpty());
+      } else {
         // Rewrite transformed records into the expected target schema
-        avroRDDOptional = transformed.map(t -> getTransformedRDD(t, reconcileSchema, schemaProvider.getTargetSchema()));
+        avroRDDOptional = transformed.map(
+            rowDataset -> getTransformedRDD(rowDataset, reconcileSchema, schemaProvider.getTargetSchema()));
+        noNewData = (!avroRDDOptional.isPresent()) || (avroRDDOptional.get().isEmpty());
       }
     } else {
-      // Pull the data from the source & prepare the write
-      InputBatch<JavaRDD<GenericRecord>> dataAndCheckpoint =
-          formatAdapter.fetchNewDataInAvroFormat(resumeCheckpointStr, cfg.sourceLimit);
-      avroRDDOptional = dataAndCheckpoint.getBatch();
-      checkpointStr = dataAndCheckpoint.getCheckpointForNextBatch();
-      schemaProvider = dataAndCheckpoint.getSchemaProvider();
+      if (recordType == HoodieRecordType.SPARK) {
+        // Pull the data from the source in row format for bulk inserts & prepare the write
```

Comment:
can you file a jira to introduce proper abstractions on this. we should unify/generify the two diff types ( JavaRdd<GenericRecord> and Dataset<Row>